### DISCUSSION TITLE
Every time an elements pushed to an array, the other elements also rerender

### DISCUSSION THREAD

**Discussion by songkeys** [pmndrs/valtio#827] [Q&A]
As the title says, what should I do to prevent the others from rerendering? Or maybe I am misusing something?

Please see this [CodeSandbox](https://codesandbox.io/p/devbox/zealous-wozniak-tgltp4?layout=%257B%2522sidebarPanel%2522%253A%2522EXPLORER%2522%252C%2522rootPanelGroup%2522%253A%257B%2522direction%2522%253A%2522horizontal%2522%252C%2522contentType%2522%253A%2522UNKNOWN%2522%252C%2522type%2522%253A%2522PANEL_GROUP%2522%252C%2522id%2522%253A%2522ROOT_LAYOUT%2522%252C%2522panels%2522%253A%255B%257B%2522type%2522%253A%2522PANEL_GROUP%2522%252C%2522contentType%2522%253A%2522UNKNOWN%2522%252C%2522direction%2522%253A%2522vertical%2522%252C%2522id%2522%253A%2522clpc1226a00083b6ghgpw23m8%2522%252C%2522sizes%2522%253A%255B70%252C30%255D%252C%2522panels%2522%253A%255B%257B%2522type%2522%253A%2522PANEL_GROUP%2522%252C%2522contentType%2522%253A%2522EDITOR%2522%252C%2522direction%2522%253A%2522horizontal%2522%252C%2522id%2522%253A%2522EDITOR%2522%252C%2522panels%2522%253A%255B%257B%2522type%2522%253A%2522PANEL%2522%252C%2522contentType%2522%253A%2522EDITOR%2522%252C%2522id%2522%253A%2522clpc1226900033b6g0maeh0sj%2522%257D%255D%257D%252C%257B%2522type%2522%253A%2522PANEL_GROUP%2522%252C%2522contentType%2522%253A%2522SHELLS%2522%252C%2522direction%2522%253A%2522horizontal%2522%252C%2522id%2522%253A%2522SHELLS%2522%252C%2522panels%2522%253A%255B%257B%2522type%2522%253A%2522PANEL%2522%252C%2522contentType%2522%253A%2522SHELLS%2522%252C%2522id%2522%253A%2522clpc1226900053b6gdzw53svw%2522%257D%255D%252C%2522sizes%2522%253A%255B100%255D%257D%255D%257D%252C%257B%2522type%2522%253A%2522PANEL_GROUP%2522%252C%2522contentType%2522%253A%2522DEVTOOLS%2522%252C%2522direction%2522%253A%2522vertical%2522%252C%2522id%2522%253A%2522DEVTOOLS%2522%252C%2522panels%2522%253A%255B%257B%2522type%2522%253A%2522PANEL%2522%252C%2522contentType%2522%253A%2522DEVTOOLS%2522%252C%2522id%2522%253A%2522clpc1226900073b6gi9okdito%2522%257D%255D%252C%2522sizes%2522%253A%255B100%255D%257D%255D%252C%2522sizes%2522%253A%255B50%252C50%255D%257D%252C%2522tabbedPanels%2522%253A%257B%2522clpc1226900033b6g0maeh0sj%2522%253A%257B%2522id%2522%253A%2522clpc1226900033b6g0maeh0sj%2522%252C%2522tabs%2522%253A%255B%255D%257D%252C%2522clpc1226900073b6gi9okdito%2522%253A%257B%2522tabs%2522%253A%255B%257B%2522id%2522%253A%2522clpc1226900063b6g86sndbld%2522%252C%2522mode%2522%253A%2522permanent%2522%252C%2522type%2522%253A%2522TASK_PORT%2522%252C%2522taskId%2522%253A%2522dev%2522%252C%2522port%2522%253A5173%252C%2522path%2522%253A%2522%252F%2522%257D%255D%252C%2522id%2522%253A%2522clpc1226900073b6gi9okdito%2522%252C%2522activeTabId%2522%253A%2522clpc1226900063b6g86sndbld%2522%257D%252C%2522clpc1226900053b6gdzw53svw%2522%253A%257B%2522id%2522%253A%2522clpc1226900053b6gdzw53svw%2522%252C%2522tabs%2522%253A%255B%257B%2522id%2522%253A%2522clpc1226900043b6gydeay5qa%2522%252C%2522mode%2522%253A%2522permanent%2522%252C%2522type%2522%253A%2522TASK_LOG%2522%252C%2522taskId%2522%253A%2522dev%2522%257D%255D%252C%2522activeTabId%2522%253A%2522clpc1226900043b6gydeay5qa%2522%257D%257D%252C%2522showDevtools%2522%253Atrue%252C%2522showShells%2522%253Atrue%252C%2522showSidebar%2522%253Atrue%252C%2522sidebarPanelSize%2522%253A15%257D)

**Comment by dai-shi** [comment]
A list of items is a tricky use case, but in your case, props have just a string, so `memo` would just work.

**Reply by songkeys** [reply]
>  but in your case, props have just a string, so memo would just work.

Ah, this code sandbox is just for a demo case. In the real world scenario, it's definitely an array of objects. Do you have any idea how to do it in this case then?

**Reply by dai-shi** [reply]
It's better to pass a proxy object instead of a snapshot object and `useSnapshot` in `Item` component. That should be safe with `memo`.

**Reply by songkeys** [reply]
> It's better to pass a proxy object instead of a snapshot object and useSnapshot in Item component.

That's what I did in my example (because of the string item, I didn't use `useSnapshot` to render, but it's the same thing - they'll still get re-rendered).

> That should be safe with memo.

So the only way to do it is still to `memo` items with the same props, I guess? This is the least preferable option for me because comparing complex objects using `JSON.stringify` hurts performance in another way.

But I understand how challenging it is to optimize a complex list in React. Thank you for the guidance anyway!
